src/Hocon/HoconObject.cs

```diff
@@ -0,0 +1,429 @@
+﻿// -----------------------------------------------------------------------
+// <copyright file="HoconImmutableObject.cs" company="Akka.NET Project">
+//      Copyright (C) 2013 - 2020 .NET Foundation <https://github.com/akkadotnet/hocon>
+// </copyright>
+// -----------------------------------------------------------------------
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Numerics;
+using System.Text;
+
+namespace Hocon
+{
+    public class HoconObject :
```

Comment:
So why this class - what's the motivation for separating all of these into their new structures? Need to understand the high level point of view.

Reply:
The main idea is to simplify the data passed back to the user. Rewriting/restructuring the existing data structure and making sure that it works with the hocon parser, substitution and external file include callback would take way too long to do and way too complicated, so instead of modifying what we have right now, I chose to transfer the tree structure to a more lean and immutable data structure.

- All of the tree nodes are based on an abstract class called `HoconElement` that will give access to all of the data access API surface.
- The `HoconObject` is modeled after the API of the standard `ImmutableSortedDictionary`
- The `HoconArray` is modeled after the API of the standard `ImmutableList`
- The `HoconLiteral` is just a simple immutable class with one value property.

The advantage is that, since everything is immutable, it is safe to pass them around, even making multiple reference to a single instance, without causing any concurrency problems, ie. it is inherently thread safe. The disadvantage is that, all the previously allocated objects are discarded, so the whole solution have a higher memory footprint than previously.

To build the tree structure, I opted to use the builder pattern. There are 3 builders:
- `HoconObjectBuilder` basically have the exact same API as `SortedDictionary<string, HoconElement>`. You `Add(key, value)` items into it and then call the `Build()` function to create an immutable `HoconObject`
- `HoconArrayBuilder` basically have the exact same API as `List<HoconElement>`. Same as `HoconObjectBuilder`, you `Add()` or `AddRange()` `HoconElement`s and then call `Build()` and it'll return a new immutable `HoconArray`.
- `HoconLiteralBuilder` basically have the exact same API as `StringBuilder`. Just add strings into it and call `Build()`.

All 3 builders are chainable in fluent way, so you can do things like `new HoconObjectBuilder(existingObjectDictionary).Add(element1).Add(element2).Build()